mlflow/genai/scorers/builtin_scorers.py

```diff
@@ -2260,6 +2285,256 @@ def instructions(self) -> str:
+        "in earlier conversation turns without forgetting, contradicting, or distorting it."
+    )
+
+    def __init__(self, /, last_turn_scorer: Scorer | None = None, **kwargs):
+        """
+        Initialize KnowledgeRetention scorer.
+
+        Args:
+            last_turn_scorer: Optional custom scorer to use for evaluating knowledge
+                retention of each turn.
+            **kwargs: Additional arguments passed to parent (name, model, etc.)
+        """
+        if last_turn_scorer is not None:
+            kwargs["last_turn_scorer"] = last_turn_scorer
+        super().__init__(**kwargs)
+
+    def _create_judge(self) -> InstructionsJudge:
+        """
+        This method is required by BuiltInSessionLevelScorer but is not used.
+        KnowledgeRetention uses composition (delegating to last_turn_scorer)
+        rather than creating its own judge.
+        """
+        raise NotImplementedError(
+            "KnowledgeRetention uses composition with last_turn_scorer "
+            "and does not use a judge directly."
+        )
+
+    @property
+    def instructions(self) -> str:
+        """
+        This property is required by BuiltInSessionLevelScorer but is not used.
+        KnowledgeRetention uses composition (delegating to last_turn_scorer)
+        rather than using its own instructions.
+        """
+        raise NotImplementedError(
+            "KnowledgeRetention uses composition with last_turn_scorer "
+            "and does not use instructions directly."
+        )
+
+    def __call__(
+        self,
+        *,
+        session: list[Trace] | None = None,
+        expectations: dict[str, Any] | None = None,
+        **kwargs,
+    ) -> Feedback:
+        """
+        Evaluate knowledge retention across conversation turns.
+
+        Args:
+            session: List of traces from the same conversation session.
+            expectations: Not used for this scorer.
+            **kwargs: Additional arguments (will raise TypeError if provided).
+
+        Returns:
+            A single Feedback object with value "yes" or "no", plus detailed rationale
+            describing which turns (if any) had retention issues.
+        """
+        if kwargs:
+            invalid_args = ", ".join(f"'{k}'" for k in kwargs.keys())
+            raise TypeError(
+                f"Session level scorers can only accept the `session` and `expectations` "
+                f"parameters. Got unexpected keyword argument(s): {invalid_args}"
+            )
+
+        if not session:
+            return Feedback(
+                name=self.name,
+                value="yes",
+                rationale="No conversation to evaluate (empty session)",
+                source=AssessmentSource(
+                    source_type=AssessmentSourceType.LLM_JUDGE,
+                    source_id=self.model or get_default_model(),
+                ),
+            )
+
+        sorted_traces = sorted(session, key=lambda t: t.info.timestamp_ms)
```

Comment:
same here with `resolve_conversation_from_session`

Reply:
Why is it necessary to call `resolve_conversation_from_session`? It seems that we would need to convert it back to `Trace` anyways in order to forward to the underlying scorer (?)